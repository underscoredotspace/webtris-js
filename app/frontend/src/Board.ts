import { BoardGrid, Size } from "./types";

export default class Board {
  private grid: BoardGrid = [];
  private size: Size;

  constructor(w = 10, h = 19) {
    this.grid = new Array(h).fill(new Array(w).fill("x"));
    this.size = { w, h };
  }

  public setGrid(board: BoardGrid) {
    this.grid = board;
  }

  public copy(): BoardGrid {
    return this.grid.map((row) => row.slice());
  }

  public clearFullRows(): number {
    let lines = 0;
    let newBoardGrid: BoardGrid = [];
    for (let row of this.grid) {
      if (row.includes("x")) {
        newBoardGrid.push(row.slice());
      } else {
        newBoardGrid.unshift(new Array(this.size.w).fill("x"));
        lines++;
      }
    }
    this.grid = newBoardGrid;
    return lines;
  }
}
