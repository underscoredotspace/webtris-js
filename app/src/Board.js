import Row from '../src/Row'

export default class Board {
  constructor() {
    this.rows = []

    for (let ndx = 0; ndx<=17; ndx++) {
      this.rows[ndx] = new Row()
    }
  }

  render() {
    const rowHTML = this.rows.map(row => row.render()).join('')
    return `<div class="board">\n\t${rowHTML}\n</div>`
  }
}