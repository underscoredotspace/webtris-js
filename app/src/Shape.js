import shapes from './shapes';

export default class Shape {
  constructor() {
    const shape = shapes[Math.floor(Math.random() * shapes.length)]

    this.type = shape.type
    this.grid = shape.grid
    this.pos = {x: shape.start, y:2}
    this.rotateFix = shape.rotateFix
    this.rotateFixNdx = 0
  }

  move(x) {
    this.pos.x += x
  }

  drop() {
    this.pos.y++
  }

  unDrop() {
    this.pos.y--
  }

  rotateCW() {
    const grid = this.grid
    const newGrid = [...Array(grid[0].length)].map(e => Array(grid.length))

    for (let row in grid) {
      for (let col in grid[row]) {
        newGrid[col][grid.length-row-1] = grid[row][col]
      }
  }

    this.grid = newGrid

    this.rotateFixNdx++
    if (this.rotateFixNdx >= this.rotateFix.length) {this.rotateFixNdx = 0}

    this.pos.x += this.rotateFix[this.rotateFixNdx].x
    this.pos.y += this.rotateFix[this.rotateFixNdx].y
  }

  rotateCCW() {
    const grid = this.grid
    const newGrid = [...Array(grid[0].length)].map(e => Array(grid.length))

    for (let row in grid) {
      for (let col in grid[row]) {
        newGrid[grid[0].length-col-1][row] = grid[row][col]
      }
    }

    this.grid = newGrid

    this.pos.x -= this.rotateFix[this.rotateFixNdx].x
    this.pos.y -= this.rotateFix[this.rotateFixNdx].y

    this.rotateFixNdx--
    if (this.rotateFixNdx < 0) {this.rotateFixNdx = this.rotateFix.length-1}
  }
}