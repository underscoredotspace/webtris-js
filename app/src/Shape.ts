import shapes from "./shapes";
import { Vector } from "./types";

export default class Shape {
  public type: string;
  public pos: Vector;
  public grid: number[][];
  private rotateFix: Vector[];
  private rotateFixNdx: number = 0;

  constructor() {
    const shape = shapes[Math.floor(Math.random() * shapes.length)];

    this.type = shape.type;
    this.grid = shape.grid;
    this.pos = { x: shape.start, y: 1 };
    this.rotateFix = shape.rotateFix;
  }

  public move(x: Vector["x"]) {
    this.pos.x += x;
  }

  public drop() {
    this.pos.y++;
  }

  public unDrop() {
    this.pos.y--;
  }

  public rotateCW() {
    if (this.type == "o") {
      return;
    }

    const grid = this.grid;
    const newGrid = [...Array(grid[0].length)].map((e) => Array(grid.length));

    grid.forEach((row, rowIndex) => {
      row.forEach((_col, colIndex) => {
        newGrid[colIndex][grid.length - rowIndex - 1] =
          grid[rowIndex][colIndex];
      });
    });

    this.grid = newGrid;

    this.rotateFixNdx++;
    if (this.rotateFixNdx >= this.rotateFix.length) {
      this.rotateFixNdx = 0;
    }

    this.pos.x += this.rotateFix[this.rotateFixNdx].x;
    this.pos.y += this.rotateFix[this.rotateFixNdx].y;
  }

  public rotateCCW() {
    if (this.type == "o") {
      return;
    }

    const grid = this.grid;
    const newGrid = [...Array(grid[0].length)].map((e) => Array(grid.length));

    grid.forEach((row, rowIndex) => {
      row.forEach((_col, colIndex) => {
        newGrid[grid[0].length - colIndex - 1][rowIndex] =
          grid[rowIndex][colIndex];
      });
    });

    this.grid = newGrid;

    this.pos.x -= this.rotateFix[this.rotateFixNdx].x;
    this.pos.y -= this.rotateFix[this.rotateFixNdx].y;

    this.rotateFixNdx--;
    if (this.rotateFixNdx < 0) {
      this.rotateFixNdx = this.rotateFix.length - 1;
    }
  }

  public render() {
    const boardHTML = ['<div class="board mini">'];
    for (let row of this.grid) {
      boardHTML.push('\t<div class="row">\n');

      for (let block of row) {
        boardHTML.push(
          `\t\t<div class="block" type="${block == 1 ? this.type : "x"}"></div>\n`,
        );
      }

      boardHTML.push("\t</div>\n");
    }

    boardHTML.push("</div>");

    return boardHTML.join("");
  }
}
