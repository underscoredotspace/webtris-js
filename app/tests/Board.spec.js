import Board from '../src/Board'
jest.mock('../src/shapes')
import Shape from '../src/Shape'

const random = (min, max) => Math.floor(Math.random() * (max - min + 1) + min)

describe('Board', () => {
  let board

  beforeEach(() => {
    board = new Board()
  })

  test('has 19 rows with 10 columns filled with 0', () => {
    expect.assertions(2)
    expect(board.board).toBeInstanceOf(Array)
    expect(board.board).toHaveLength(190)
  })

  test('render method', () => {
    expect.assertions(1)
    expect(board.render()).toMatchSnapshot()
  })

  test('newShape method should add new Shape to this.shape', () => {
    board.newShape()
    expect(board.shape).toBeInstanceOf(Shape)
  })

  test('row method returns row', () => {
    const expected = new Array(10).fill('x')
    expect(board.row(0)).toEqual(expected)
  })

  test('render method with board.shape', () => {
    board.newShape()
    expect(board.render()).toMatchSnapshot()
  })

  test('update method where shape is not at bottom', () => {
    board.newShape()
    board.update()
    expect(board.render()).toMatchSnapshot()
  })
})