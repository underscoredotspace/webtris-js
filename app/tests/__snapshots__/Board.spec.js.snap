// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board render method 1`] = `
"<div class=\\"board\\">	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
</div>"
`;

exports[`Board render method with board.shape 1`] = `
"<div class=\\"board\\">	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"l\\"></div>
		<div class=\\"block\\" type=\\"l\\"></div>
		<div class=\\"block\\" type=\\"l\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"l\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
</div>"
`;
